### PR TITLE
simplify bank drop calls

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem
The extra parameter, *is_from_abs* in `purge_slot` function is only used for assert. It seems unnecessary. Furthermore, it is also a bit confusing since caller from block_store also set it to true.

*is_bank_drop_callback_enabled* flag seems also unnecessary since the call back is an Option. Besides, this flag is set in a callback create function, which seems not to be the right place. The right place should be at the caller, which installed the drop callback to the root bank. 

#### Summary of Changes

simplify the bank drop calls.

Fixes #

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -6754,9 +6754,7 @@ impl Drop for Bank {
             drop_callback.callback(self);
         } else {
```
Comment (HaoranYi) [inline runtime/src/bank.rs:6755]: @brooksprumo

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -1374,7 +1374,7 @@ impl Bank {
             ),
             transaction_log_collector: Arc::<RwLock<TransactionLogCollector>>::default(),
             feature_set: Arc::<FeatureSet>::default(),
-            drop_callback: RwLock::<OptionalDropCallback>::default(),
+            drop_callback: RwLock::new(OptionalDropCallback(None)),
```
Comment (brooksprumo) [inline runtime/src/bank.rs:1377]: 👍

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -6754,9 +6754,7 @@ impl Drop for Bank {
             drop_callback.callback(self);
         } else {
             // Default case for tests
-            self.rc
-                .accounts
-                .purge_slot(self.slot(), self.bank_id(), false);
+            self.rc.accounts.purge_slot(self.slot(), self.bank_id());
```
Comment (brooksprumo) [inline runtime/src/bank.rs:6757]:
```suggestion
            self.rc.accounts.accounts_db.purge_slot(self.slot(), self.bank_id());
```

**Review by brooksprumo — COMMENTED** [review]
Ok, I'm on board with this change now. I think it's now safe since #23999 sets the bank drop callback on load, before any new banks are created or any bank refs are passed out.

I had some nits below, but otherwise looks good. Once the changes are in and the CI passes coverage, please re-request a review. Thanks!

**Review by brooksprumo — DISMISSED** [review]
This looks good to me. I think it'd probably be worthwhile to have another person approve as well; to make sure I didn't miss anything.

**Comment by HaoranYi** [comment]
@carllin @jeffwashington Can you help to review this? Thanks!

**Comment by jeffwashington** [comment]
@carllin I think this needs to be you.

**Comment by carllin** [comment]
The flag serves a few purposes I think which are still useful:

1. Documenting that the function needs to be serialized with all the other background account processes
2. Forcing anybody calling this code to reason about whether their function is always serialized with said background processes
3.  If in step 2, the developer makes a mistake in their reasoning, the code will panic instead of silently corrupting the state. For instance, this was useful to catch that this code: https://github.com/solana-labs/solana/blob/781094edb2a8a9a905be128353eba5437ff2368d/runtime/src/bank.rs#L6809 could run concurrently with `ABS` on startup

That being said, the `is_abs` flag could maybe be more clearly renamed to `is_serialized_with_abs` to better convey its purpose

**Comment by HaoranYi** [comment]
OK. Close the pull request. Let's leave the code as it is.

**Comment by HaoranYi** [comment]
Ideally, the check shouldn't be on the purge_slot fn. It should be inside the drop_callback. 
I can imagine that we add a bool flag into the bank to indicates that we are 'dropping' then have the drop_callback to assert that flag. However, this is difficult because we can't change the bank struct. I wonder if there is any unused fields in the bank that can be repurposed for using as the "dropping" flag.

**Comment by brooksprumo** [comment]
> However, this is difficult because we can't change the bank struct.

Why can't we change the Bank struct?

**Comment by HaoranYi** [comment]
https://github.com/solana-labs/solana/blob/2e5042d8bd707526ab44912111226defe0130d2a/runtime/src/bank.rs#L1085
In the comments, it is saying that serialization/deserialization are implemented elsewhere. I assume that Bank is embedded in some other serialized struct?

**Comment by carllin** [comment]
> Ideally, the check shouldn't be on the purge_slot fn. It should be inside the drop_callback.
I can imagine that we add a bool flag into the bank to indicates that we are 'dropping' then have the drop_callback to assert that flag.

I think it makes sense to leave the flag in the `purge_slot()` function because that's the exact function that needs to be serialized with `ABS`

From my comment above: https://github.com/solana-labs/solana/pull/24142#issuecomment-1092276909, one of the purposes is `Forcing anybody calling this code to reason about whether their function is always serialized with said background processes`, which can only be enforced if the flag is in `purge_slot()`

**Comment by HaoranYi** [comment]
what do you mean by `always serialized with said background processes`?

**Comment by carllin** [comment]
It means that the `purge_slot()` function and all the functions in `AccountsBackgroundService` cannot run concurrently, they must run in a serial fashion

**Review by carllin — APPROVED** [review]

**Pull request merged**